sstables/kl/reader_impl.hh

```diff
@@ -170,277 +144,248 @@ private:
+                } else {
+                    ret = _consumer.consume_cell(to_bytes_view(_key),
+                            fragmented_temporary_buffer::view(_val_fragmented), _u64, _ttl, _expiration);
+                }
+                // after calling the consume function, we can release the
+                // buffers we held for it.
+                _key.release();
+                _val_fragmented.remove_prefix(_val_fragmented.size_bytes());
+                if (ret == row_consumer::proceed::no) {
+                    co_await wait();
+                    if (should_end) {
+                        co_return;
+                    }
+                }
             }
-            break;
         }
-        case state::STOP_THEN_ATOM_START:
-            _state = state::ATOM_START;
-            return row_consumer::proceed::no;
+    }
+    future<> producer(bool skip_row) {
+        return producer_main(skip_row).handle_exception_type([this] (malformed_sstable_exception& exp) {
+            throw malformed_sstable_exception(exp.what(), _sst->get_filename());
+        }).handle_exception([this] (auto eptr) {
+            return _production_sem.broken(eptr);
+        });
+    }
+    void end_production() {
+        if (!_production.available()) {
+            *_production_end = true;
+            _production_end = nullptr;
+            _production_sem.signal();
         }
-
-        return row_consumer::proceed::yes;
     }
 public:
-
     data_consume_rows_context(const schema&,
                               const shared_sstable sst,
                               row_consumer& consumer,
                               input_stream<char>&& input, uint64_t start, uint64_t maxlen)
-                : continuous_data_consumer(consumer.permit(), std::move(input), start, maxlen)
+                : async_data_consumer(consumer.permit(), std::move(input), start, maxlen)
                 , _consumer(consumer)
                 , _sst(std::move(sst))
+                , _production_sem(semaphore_type(0))
+                , _production(producer(false))
     {}
+    ~data_consume_rows_context() {
+        end_production();
+    }
 
-    void verify_end_state() {
-        // If reading a partial row (i.e., when we have a clustering row
-        // filter and using a promoted index), we may be in ATOM_START or ATOM_START_2
-        // state instead of ROW_START. In that case we did not read the
-        // end-of-row marker and consume_row_end() was never called.
-        if (_state == state::ATOM_START || _state == state::ATOM_START_2) {
-            _consumer.consume_row_end();
-            return;
-        }
-        if (_state != state::ROW_START || primitive_consumer::active()) {
-            throw malformed_sstable_exception("end of input, but not end of row");
-        }
+    future<> await_production() {
+        _production_sem.signal();
+        co_return co_await _production_sem.wait();
```

Comment:
It might be safer to make two semaphores one for data_consume_rows_context -> producer_main communication and another for producer_main->data_consume_rows_context. Just to make sure there are no weird races between them.

Reply:
We want to reduce the number of semaphores, not increase them.

Reply:
(there should be no races in this code, since there is no real concurrency)

Reply:
What about producer_main coroutine being started in the constructor of data_consume_rows_context and fill_buffer being called before that coroutine reaches _production_sem.wait()?

Reply:
Maybe that's impossible - I really don't have any intuition about coroutines.

Reply:
I'm not saying if it's possible or not, I'm saying that the logical path has no concurrency. We read data from disk, wait for it, parse in memory, wait if the buffer is full, resume, parse some more, run out of data, read from disk, rinse and repeat. Nothing happens in parallel. So if there are races, this code introduces them. Rather than fixing with an expensive semaphore, it should not introduce them in the first place.

Reply:
Fair enough.

Reply:
Couldn't we convert all this to a pull model with `co_yield`? All the complication rises from the currently used push model, where the reader starts a fiber, which pushes fragments into the reader's buffer, creating a circle (R=reader, P=parser):

reader (R) -> read() (P) -> push fragment (P) -> is buffer full (R) -> yes ? stop/continue (P)

With a generator we could have a simple loop in the reader's `fill_buffer()` (fantasy pseudo code follows):

```C++
while (!_is_buffer_full) {
    push_mutation_fragment(co_await _parser.get_next_fragment());
}
```
Internally, `get_next_fragment()`, uses `co_yield` to return the last parsed fragment and to suspend its state until the next fragment is requested.

Reply:
I think co_yield will be inefficient because it would context switch on each fragment

Reply:
Is it really a context switch though? You are basically just calling a function for each fragment.

Reply:
I think the cost is minor. As long as we are not yielding to the consumer (and just buffering) there are few extra instructions added. Now that perf_fast_forward reports instructions per fragment this can be easily measured.

Reply:
If co_yield buffers, then the context switch is amortized. 

It still seems problematic. If the parser buffers, it has to over-parse. The reader could stop earlier.

I don't think it simplifies things. I see how it makes things more complicated. The parser knows the type of the fragment. If we produce mutation_fragment, this information is erased and the reader has to figure it out again. The same logic which would live in the parser's handler will also have to live in the reader's loop.

Reply:
> If co_yield buffers, then the context switch is amortized.
> 
> It still seems problematic. If the parser buffers, it has to over-parse. The reader could stop earlier.
> 

The idea is to have the same behavior as now in terms of buffering, just with a nicer interface.

`data_consume_rows_context_m::_state` is replaced with the coroutine remembering the instruction pointer across suspends.
switch/goto are replaced with normal nested loops
`co_yield ...` calls `operator co_await`, and on the return object of that it calls `await_ready()`. We can arrange for that to return false if no preemption needs to happen and if the buffer is not full; this should translate directly to the existing `if` checks.


> I don't think it simplifies things. I see how it makes things more complicated. The parser knows the type of the fragment. If we produce mutation_fragment, this information is erased and the reader has to figure it out again. The same logic which would live in the parser's handler will also have to live in the reader's loop.

The information is not erased, it remains in the instruction pointer. Any local state (which is now in member variables) is also retained.